url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/fsa/logics/Utils.java

```diff
@@ -0,0 +1,391 @@
+    if (termList.size() > 0) {
+      // termList = expandDontCares(termList,0);
+      QmcFormula qmcf = new QmcFormula(termList);
+      qmcf.reduceToPrimeImplicants();
+      qmcf.reducePrimeImplicantsToSubset();
+      simplifiedTf = termsListToTf(qmcf.termList);
+    } else
+      return SimpleTf.AcceptsNone();
+
+    return simplifiedTf;
+
+  }
+
+  // at this point I know that each term has all possible Tfs in the formula
+  private static List<Term> applyKnowledge(List<Term> termList) {
+
+	
+	Term[] at = (Term[])((ArrayList<Term>) termList).toArray(new Term[termList.size()]);
+	Set<Implication> tosubstitute = new HashSet<Implication>();
+	
+	// knowledge is A -> B,
+	// term has A and B 
+	//      A=1, B=1 in a term, then B must be set as dontcare
+	//      A=1, B=0 in a term, then the term should be removed
+	//      A=0, B=1 in a term, know nothing
+	//      A=0, B=0 in a term, know nothing
+	//
+	// 	TODO	A=2, B=1 in a term, then all occurrences of A in all the terms have to be substituted by B.
+	// 	TODO	A=2, B=0 in a term, know nothing
+	// 	TODO	A=1, B=2 in a term, know nothing
+	// 	TODO	A=0, B=2 in a term, know nothing
+	// knowledge is A -> -B,
+	// term has A and B  
+	//      A=1, B=1 in a term, then the term should be removed
+	//      A=1, B=0 in a term, then B must be set as dontcare
+	//      A=0, B=1 in a term, then A must be set as dontcare
+	//      A=0, B=0 in a term, know nothing
+	//
+	// 	TODO	A=2, B=1 in a term, 
+	// 	TODO	A=2, B=0 in a term, 
+	// 	TODO	A=1, B=2 in a term, 
+	// 	TODO	A=0, B=2 in a term, 
+	// knowledge is -A -> B,
+	// term has A and B  
+	//      A=1, B=1 in a term, know nothing
+	//      A=1, B=0 in a term, know nothing
+	//      A=0, B=1 in a term, then B must be set as dontcare
+	//      A=0, B=0 in a term, then the term should be removed
+	//
+	// TODO		A=2, B=1 in a term, 
+	// TODO		A=2, B=0 in a term, 
+	// TODO		A=1, B=2 in a term, 
+	// TODO		A=0, B=2 in a term, 
+	// knowledge is -A -> -B,
+	// term has A and B 
+	//      A=1, B=1 in a term, know nothing
+	//      A=1, B=0 in a term, know nothing
+	//      A=0, B=1 in a term, then the term should be removed
+	//      A=0, B=0 in a term, then B must be set as dontcare
+	//
+	// 	TODO	A=2, B=1 in a term, 
+	// 	TODO	A=2, B=0 in a term, 
+	// 	TODO	A=1, B=2 in a term, 
+	// 	TODO	A=0, B=2 in a term, 
+
+	
+	
+	for (Implication impl : Knowledge.implications) {
+	  // knowledge is A -> B,
+	  if (!impl.all.isNot() && !impl.are.isNot()) {
+		for (int i = 0; i < at.length; i++) {
+		  if (at[i] != null) {
+			TfTerm all = null;
+			TfTerm are = null;
+			for (TfTerm tft : at[i].varVals) {
+			  if (impl.all.equals(tft.tf) && tft.b != TfTerm.DontCare.b)
+				all = tft;
+			  if (impl.are.equals(tft.tf) && tft.b != TfTerm.DontCare.b)
+				are = tft;
+			}
+			if (all != null && are != null) {
+			  if (all.b == 1 && are.b == 1)
+				are.b = TfTerm.DontCare.b;
+			  if (all.b == 1 && are.b == 0)
+				at[i] = null;
+			}
+		  }
+		}
+	  }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 138:157, 180:199.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.